### ISSUE TITLE
Installing MbedTLS fails on Linux with glibc-2.12

### ISSUE THREAD

**Issue by Alexander-Barth** [JuliaWeb/MbedTLS.jl#170]
I get the following error when I build  MbedTLS with Julia 1.0.0 a Scientific Linux 6.6 system:

```
(v1.0) pkg> build MbedTLS
  Building MbedTLS → `~/.julia/packages/MbedTLS/McYwM/deps/build.log`
┌ Error: Error building `MbedTLS`: 
│ ┌ Warning: platform_key() is deprecated, use platform_key_abi() from now on
│ │   caller = ip:0x0
│ └ @ Core :-1
│ [ Info: using prebuilt binaries
│ ERROR: LoadError: LibraryProduct(nothing, ["libmbedcrypto"], :libmbedcrypto, "Prefix(/home/ulg/gher/abarth/.julia/packages/MbedTLS/McYwM/deps/usr)") is not satisfied, cannot generate deps.jl!
│ Stacktrace:
│  [1] error(::String) at ./error.jl:33
│  [2] #write_deps_file#151(::Bool, ::Function, ::String, ::Array{LibraryProduct,1}) at /home/ulg/gher/abarth/.julia/packages/BinaryProvider/zp4zb/src/Products.jl:408
│  [3] (::getfield(BinaryProvider, Symbol("#kw##write_deps_file")))(::NamedTuple{(:verbose,),Tuple{Bool}}, ::typeof(write_deps_file), ::String, ::Array{LibraryProduct,1}) at ./none:0
│  [4] top-level scope at none:0
│  [5] include at ./boot.jl:317 [inlined]
│  [6] include_relative(::Module, ::String) at ./loading.jl:1038
│  [7] include(::Module, ::String) at ./sysimg.jl:29
│  [8] include(::String) at ./client.jl:388
│  [9] top-level scope at none:0
│ in expression starting at /home/ulg/gher/abarth/.julia/packages/MbedTLS/McYwM/deps/build.jl:67
└ @ Pkg.Operations /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Pkg/src/Operations.jl:1068

```

It seems that the glibc version is too old. Was it an intentional choice to drop these Linux versions (based on RedHat 6) ?

```
dlopen("/home/ulg/gher/abarth/.julia/packages/MbedTLS/McYwM/deps/usr/lib/libmbedcrypto.so")
ERROR: could not load library "/home/ulg/gher/abarth/.julia/packages/MbedTLS/McYwM/deps/usr/lib/libmbedcrypto.so"
/lib64/libc.so.6: version `GLIBC_2.14' not found (required by /home/ulg/gher/abarth/.julia/packages/MbedTLS/McYwM/deps/usr/lib/libmbedcrypto.so)
```

**Comment by quinnj** [comment]
@staticfloat, do you understand better if we intended to drop glibc support for older versions w/ the most recent MbedTLSBuilder bump?

**Comment by staticfloat** [comment]
Yes, it was intentional.  It is very difficult to support older glibc versions because there are build system bugs that have been fixed in later releases and that are not easily backportable.  We bumped up to 2.17 so RHEL7 (e.g. Scientific Linux 7) should work.

I understand that in many cases "just upgrade" is not an easy ask.  I will re-investigate the backports necessary to get `2.12.2` building and passing our test suites, and if possible I will push the glibc requirements back to that for the `x86_64` and `i686` builds.  If I can do this with reasonable effort, it will be included in the next published rootfs, which means Jacob will just need to trigger a new build and things should "just work".

The official binaries of Julia itself on `x86_64` require 2.12.2, and on `i686` require 2.17 (because of bugs in old glibc versions on the i686 architecture).  It is technically possible to build Julia on a system old enough that it uses even older versions, and that will cause similar problems to this.  We definitely won't ever be able to support anything older than 2.12.2, and unfortunately most of our test cases and test machines are new enough that I didn't notice this discrepancy between the official binary support (2.12.2 on `x86_64`) and BinaryBuilder.

We will always strive to be consistent in our support bounds, and so for the 1.0 support window (which is very long) these glibcs should be safe.  But don't expect future releases to be able to support 2.12.2; eventually we're going to need to upgrade and once the official Julia binaries start getting built on a system that only supports back to 2.17, for example, the BB minimum supported targets will probably move with it, unless there is a very strong reason not to use the newer, more stable releases.

**Comment by quinnj** [comment]
We should still be able to hit the "compile from source" [fallback](https://github.com/JuliaWeb/MbedTLS.jl/blob/master/deps/build.jl#L50) though right? How do we ensure BP doesn't use prebuilt binaries in this old glibc case so it falls back to compiling? (or using julia-shipped binaries)

**Comment by staticfloat** [comment]
There are two ways to do this: before trying to install BP binaries at all, we can try to figure out the current glibc version and gate installation based on that.  Alternatively, we could try installing them, and if we're still unsatisfied after installing them, we can fall back to source build.

Right now, your build.jl checks to see if a tarball is available and installs it, so we should add a further check to see if it's satisfied and if not trigger the compile-from-source fallback.  That would look like this:

```julia
    if haskey(download_info, platform_key()) && get(ENV, "FORCE_BUILD", "false") != "true" && !haskey(ENV, "USE_GPL_MBEDTLS")
        # Download and install binaries
        url, tarball_hash = download_info[platform_key()]
        install(url, tarball_hash; prefix=prefix, force=true, verbose=verbose)
        @info "using prebuilt binaries"
    end

    if all(satisfied(p; verbose=verbose) for p in juliaproducts) && get(ENV, "FORCE_BUILD", "false") != "true"
        @info "using julia-shippied binaries or successfully downloaded BinaryProvider binaries"
        products = juliaproducts
    else
        @info "attempting source build"
        VERSION = "2.13.0"
        url, hash = haskey(ENV, "USE_GPL_MBEDTLS") ?
            ("https://tls.mbed.org/download/mbedtls-$VERSION-gpl.tgz", "a08ddf08aae55fc4f48fbc6281fcb08bc5c53ed53ffd15355ee0d75ec32b53ae") :
            ("https://tls.mbed.org/download/mbedtls-$VERSION-apache.tgz", "593b4e4d2e1629fc407ab4750d69fa309a0ddb66565dc3deb5b60eddbdeb06da")
        download_verify(url, hash, joinpath(@__DIR__, "mbedtls.tgz"), force=true, verbose=true)
        unpack(joinpath(@__DIR__, "mbedtls.tgz"), @__DIR__; verbose=true)
        withenv("VERSION"=>VERSION) do
            run(Cmd(`./build.sh`, dir=@__DIR__))
        end
        if any(!satisfied(p; verbose=verbose) for p in products)
            error("attempted to build mbedtls shared libraries, but they couldn't be located (deps/usr/lib)")
        end
    end
```

**Comment by staticfloat** [comment]
Okay getting `2.12.2` to work against wasn't so bad; my fixes for other parts of glibc had already done a good chunk of the work.  The newest BinaryBuilder RootFS supports `2.12.2` on `x86_64` and `i686` again, so @quinnj feel free to run your builder repository again, and the generated tarballs should be linked for maximum compatibility.

**Comment by quinnj** [comment]
Building: https://github.com/JuliaWeb/MbedTLSBuilder/releases/tag/v0.16.0

**Comment by quinnj** [comment]
Should be fixed by #171

**Comment by Alexander-Barth** [comment]
Thank you very much! I confirm that it works now.
